ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,464 @@
+2.  The user will want to manually create **additional** [Environments] (for example, a prod environment). The user may want to use our compute resources provided in the form of a new **namespace on a devsandbox member cluster** for this ([STONE-183](https://issues.redhat.com/browse/STONE-183)) or they may want to **bring their own cluster** as a target ([STONE-162](https://issues.redhat.com/browse/STONE-162)).
+3.  The integration-service expects to be able to create **ephemeral** [Environments] for automated testing purposes ([STONE-114](https://issues.redhat.com/browse/STONE-114)). For our short-term goals, the automated testing use case requires the same kind of compute as for the dev and stage [Environments] (devsandbox member cluster namespaces), but will expand to include other kinds of deployment targets in the future - like hypershift clusters ([STONE-185](https://issues.redhat.com/browse/STONE-185)).
+
+## Decision
+
+We are going split the two original purposes of the [Environment] CR into different APIs.
+
+- The [Environment] will still represent the first purpose of requesting that HAS and the GitOps service **recognize a new deployment destination**.
+- For the second purpose of **provisioning a new deployment target** we are going to introduce a new [DeploymentTarget] and [DeploymentTargetClaim] API, which will be referenced from an [Environment].
+
+The new [DeploymentTarget] API is designed to emulate storage management APIs (see [persistent-volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) and [storage provisioner design](https://github.com/kubernetes-csi/external-provisioner/blob/master/doc/design.md) for reference).
+
+### CRDs
+
+#### DeploymentTarget (DT)
+
+A deployment target, usually a K8s api endpoint. The credentials for connecting to the target will
+be stored in a secret which will be referenced in the clusterCredentialsSecret field. A [DT] Can be
+created manually by a user, or dynamically using a provisioner. The `phase` section shows the lifecycle phase of the [DT].
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTarget
+metadata:
+    name: prod-dt
+spec:
+    deploymentTargetClassName: isolation-level-namespace
+    kubernetesCredentials:
+        defaultNamespace: team-a--prod-dtc
+        apiURL: …
+        clusterCredentialsSecret: team-a--prod-dtc--secret
+    claimRef:
+        name: prod-dtc
+status:
+    phase: Bound
+
+```
+
+Phases:
+
+**Pending:** DT is not yet available for binding.
+
+**Available:** DT waits for a Claim to be bound to.
+
+**Bound:** The DT was bounded to a DTC.
+
+**Released:** The DT was previously bound to a DTC which got deleted. external resources were not freed.
+
+**Failed:** DT was released from its claim, but there was a failure during the release of external resources.
```

Comment:
Is the failure to release the only case that will cause it in the failed phase? What about the failed case where DT failed during creation in the first place, e.g. secret is incorrect?

Reply:
I think in that case, the DT should not have been created by the provisioner at all. Instead, the provisioner should emit an `Event` complaining that it cannot create a DT that is being requested by a DTC for some reason - it should describe the reason in the Event.